### PR TITLE
Feature/parental control

### PR DESCRIPTION
**Pull request by MasterK0927**
1) Merged codebases 
2) Implemented the url blocking

### CODE REVIEW

File: assets/rules/socialMediaBlockRules.js [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        "urlFilter": "/*skype*/",
+        "resourceTypes": ["main_frame"]
+      }
+    },
+    {
+      "id": 596,
+      "priority": 1,
+      "action": { "type": "block" },
+      "condition": {
+        "urlFilter": "/*youtube*/",
```
Comment (udaybansal19) [inline assets/rules/socialMediaBlockRules.js]: we can allow youtube. its not a social media site. 
If we can block youtube and only allow youtube kids, that would be a good to have feature

File: content/content_scripts/popup.js [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+  socialToggle.addEventListener("change", function () {
+    socialChecked = this.checked;
+  });
+
+  gamingToggle.addEventListener("change", function () {
+    gamingChecked = this.checked;
+  });
+
+  const handleServiceWorkerInjection = async () => {
```
Comment (udaybansal19) [inline content/content_scripts/popup.js]: this is not a good practice to use. what if we need to add another checkbox aur another type of rules to add. Our conditional statements will increase exponentially. 
are you still facing the issue with id?

Comment (MasterK0927) [reply content/content_scripts/popup.js]: The issue with the I'd is solved

File: assets/rules/socialMediaBlockRules.js (hunk 1 above)
Comment (MasterK0927) [reply assets/rules/socialMediaBlockRules.js]: YouTube can't be blocked

File: assets/rules/defaultBlockRules.js [hunk 3]
```diff
@@ -0,0 +1,732 @@
...
+        "urlFilter": "/*pornolienx*/",
+        "resourceTypes": ["main_frame"]
+      }
+    },
+    {
+      "id": 514,
+      "priority": 1,
+      "action": { "type": "block" },
+      "condition": {
+        "urlFilter": "/*porndor*/",
```
Comment (udaybansal19) [inline assets/rules/defaultBlockRules.js]: We can just add the check on keyword `Porn` and it should block all sites containing that string. we don't need to separately add all domains, that list would be very big

File: assets/rules/defaultBlockRules.js [hunk 4]
```diff
@@ -0,0 +1,732 @@
...
+        "urlFilter": "/*sexvideosxxx*/",
+        "resourceTypes": ["main_frame"]
+      }
+    },
+    {
+      "id": 549,
+      "priority": 1,
+      "action": { "type": "block" },
+      "condition": {
+        "urlFilter": "/*watchhindiporn2*/",
```
Comment (udaybansal19) [inline assets/rules/defaultBlockRules.js]: just rules based on keywords, not full domain names

File: assets/rules/socialMediaBlockRules.js (hunk 1 above)
Comment (udaybansal19) [reply assets/rules/socialMediaBlockRules.js]: so lets remove the rule from here. we should be able to block it though

File: content/content_scripts/popup.js (hunk 2 above)
Comment (udaybansal19) [reply content/content_scripts/popup.js]: lets use the design pattern we were using before then. This is not scalable

File: assets/rules/defaultBlockRules.js (hunk 3 above)
Comment (MasterK0927) [reply assets/rules/defaultBlockRules.js]: Yeah, although it is there, it is blocking all the domains with the keyword, but for more security added more keywords too

Comment (udaybansal19) [reply assets/rules/defaultBlockRules.js]: lets remove the extra ones which would be handled anyway. we have a limit on number of rules as well so lets add them conviously

File: assets/Logo.png
Comment (udaybansal19) [inline assets/Logo.png:1]: change this icon

File: content/content_scripts/popup.js [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+  const logoutForm = document.getElementById("logoutForm");
+  const logoutPassword = document.getElementById("logoutPassword");
+  const signInContainer = document.getElementById("signInContainer");
+  const socialToggle = document.getElementById("blockSocialMediaCheckbox");
+  const gamingToggle = document.getElementById("blockGamesCheckbox");
+
+  // Function to update popup content based on login status
+  const updatePopupContent = async () => {
+    // Check if user is logged in
+    const data = await new Promise((resolve, reject) => {
```
Comment (udaybansal19) [inline content/content_scripts/popup.js]: remove promises. use async await

File: content/content_scripts/popup.js [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        }
+      });
+    });
+
+    if (data.loggedIn) {
+      // User is logged in
+      signUpContainer.style.display = "none";
+      signInContainer.style.display = "none";
+      kidsContent.style.display = "flex";
+      kidsContent.style.flexDirection = "column";
```
Comment (udaybansal19) [inline content/content_scripts/popup.js]: can be removed

File: content/content_scripts/popup.js [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+  // Update popup content when the popup is opened
+  await updatePopupContent();
+
+  // Event listener for login form submission
+  registerForm.addEventListener("submit", async (event) => {
+    event.preventDefault();
+    const pass1 = cpasswordInput.value;
+    const pass2 = passwordInput.value;
+
+    statusDiv[0].textContent = "";
```
Comment (udaybansal19) [inline content/content_scripts/popup.js]: we can have a single statusDiv

File: content/content_scripts/popup.js [hunk 8]
```diff
@@ -0,0 +1,199 @@
+import { injectServiceWorker } from "../../service_worker/background.js";
+
+document.addEventListener("DOMContentLoaded", async () => {
```
Comment (udaybansal19) [inline content/content_scripts/popup.js]: put this into a separate function

File: content/content_scripts/popup.js (hunk 2 above)
Comment (udaybansal19) [reply content/content_scripts/popup.js]: send an array of enabled boxes to worker

File: content/ui/home.html [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+    <meta charset="UTF-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <title>Document</title>
+</head>
+<body>
+    <div id="browsingTime">
+        <h1>Ping kids mode!</h1>
+    </div>
+    <div id="browsingTimeOver" style="display: none;">
+        <h1>Your browsing time is over. Go back to study from your books.</h1>
```
Comment (udaybansal19) [inline content/ui/home.html]: improve UI for this

Comment (udaybansal19) [reply content/ui/home.html]: Improve session timeout UI

File: content/ui/home.html [hunk 10]
```diff
@@ -0,0 +1,17 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <title>Document</title>
+</head>
+<body>
+    <div id="browsingTime">
+        <h1>Ping kids mode!</h1>
```
Comment (udaybansal19) [inline content/ui/home.html]: on logging in new tab page should be shown and the browser extension icon should show that the kids mode is enabled along with a tooltip that kids mode is enabled

File: content/ui/popup.css [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+.status{
+  margin-bottom: 10px;
+}
+
+form{
+  display: flex;
+    flex-direction: column;
+}
+
+input{
```
Comment (udaybansal19) [inline content/ui/popup.css]: replace html tags with class names or Ids

File: content/ui/popup.html [hunk 12]
```diff
@@ -0,0 +1,49 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+  <meta charset="UTF-8">
+  <meta name="viewport" content="width=device-width, initial-scale=1.0">
+  <title>My Extension Popup</title>
```
Comment (udaybansal19) [inline content/ui/popup.html]: Kids mode

File: content/ui/sessionTimeout.html [hunk 13]
```diff
@@ -0,0 +1,11 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <title>Document</title>
+</head>
+<body>
+    <h1> Your session has expired. Kindly logout from kids mode and start a new session if you want</h1>
```
Comment (udaybansal19) [inline content/ui/sessionTimeout.html]: is this needed

File: service_worker/background.js [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+import {socialMediaBlockRules} from "../assets/rules/socialMediaBlockRules.js";
+import {gamingSiteRules} from "../assets/rules/gamesBlockRules.js";
+
+let startTime;
+let timerId;
+let timeoutDuration;
+const startSessionTimeout = () => {
+    startTime = Date.now();
+    timerId = setTimeout(sessionTimeout, timeoutDuration);
+    console.log(timeoutDuration)
```
Comment (udaybansal19) [inline service_worker/background.js]: console log needed in final output

File: service_worker/background.js [hunk 15]
```diff
@@ -0,0 +1,327 @@
...
+                        // Close current windows
+                        chrome.windows.getAll({ populate: true }, (windows) => {
+                            windows.forEach((window) => {
+                                chrome.windows.remove(window.id);
+                            });
+                        });
+    
+                        // Create a new window with Google
+                        chrome.windows.create({
+                            url: '../content/ui/home.html',
```
Comment (udaybansal19) [inline service_worker/background.js]: open new tab

File: service_worker/background.js [hunk 16]
```diff
@@ -0,0 +1,327 @@
...
+        });
+
+        const storedPassword = data.password;
+
+        const hash = await hashPassword(password);
+
+        if (storedPassword === hash) {
+            
+            try {
+                timeoutDuration = 1 * 60 * 1000
```
Comment (udaybansal19) [inline service_worker/background.js]: need this input from user

File: service_worker/background.js [hunk 17]
```diff
@@ -0,0 +1,327 @@
...
+
+        const storedPassword = data.password;
+
+        const hash = await hashPassword(password);
+
+        // Check if the provided password matches the stored password
+        if (storedPassword === hash) {
+            removeServiceWorker();
+            clearTimeout(timerId);
+            chrome.storage.local.set({loggedIn : false, sessionTimeout:false, timeLeft: 1 * 60 * 1000},)
```
Comment (udaybansal19) [inline service_worker/background.js]: remove hardcoded value

Comment (udaybansal19) [reply service_worker/background.js]: check if this a promise or not

File: service_worker/background.js [hunk 18]
```diff
@@ -0,0 +1,327 @@
...
+            await allowHttpsSearchAsync();
+
+            chrome.windows.getAll({ populate: true }, (windows) => {
+                windows.forEach((window) => {
+                    chrome.windows.remove(window.id);
+                });
+            });
+
+            chrome.windows.create({
+                url: 'https://google.com',
```
Comment (udaybansal19) [inline service_worker/background.js]: send this to new tab page

File: service_worker/background.js [hunk 19]
```diff
@@ -0,0 +1,327 @@
...
+    console.log(timeoutDuration)
+}
+
+// Function to update time in local storage every minute
+const updateTimeInLocalStorage = () => {
+    setInterval(() => {
+        const currentTime = Date.now();
+        // console.log('Current time:', currentTime);
+        chrome.storage.local.set({ timeLeft: timeoutDuration -  (currentTime - startTime) })
+    }, 1000); // 60000 milliseconds = 1 minute
```
Comment (udaybansal19) [inline service_worker/background.js]: should be 1 minute

File: service_worker/background.js [hunk 20]
```diff
@@ -0,0 +1,327 @@
...
+        } else {
+            sendResponse({ success: false, error: '*Invalid Credentials' });
+        }
+    } catch (error) {
+        console.error('Error logging out:', error);
+        sendResponse({ success: false, error: 'An error occurred while logging out' });
+    }
+}
+
+export const injectServiceWorker = async (socialMediaChecked, gamingChecked) => {
```
Comment (udaybansal19) [inline service_worker/background.js]: refactor this to support array of enabled checks

File: content/content_scripts/popup.js [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+
+  // Function to handle service worker injection based on toggle status and update the toggles array
+  const handleServiceWorkerInjection = async () => {
+    const checkedToggles = toggles.filter((toggle) => toggle.checked);
+    await injectServiceWorker(toggles,checkedToggles);
+  };
+
+  // Add event listeners to toggle elements
+  toggles.forEach((toggle) => {
+    toggle.element.addEventListener("change", function () {
```
Comment (udaybansal19) [inline content/content_scripts/popup.js]: we do not need an event listener if we are using the state of check box on submit

File: service_worker/background.js [hunk 22]
```diff
@@ -0,0 +1,327 @@
...
+export const injectServiceWorker = async (toggles,checkedToggles) => {
+    const rulesToInject = [];
+    const oldRules = await chrome.declarativeNetRequest.getDynamicRules();
+    const oldRulesIds = oldRules.map(rule => rule.id);
+
+    rulesToInject.push(...defaultBlockRules);
+
+    checkedToggles.forEach((toggle) => {
+        const {element}= toggle;
+        const index = toggles.findIndex((t) => t.element === element);
```
Comment (udaybansal19) [inline service_worker/background.js]: why are we adding switch case on indexes. we can just add on the element

File: content/content_scripts/popup.js [hunk 23]
```diff
@@ -0,0 +1,180 @@
...
+      }
+    } catch (error) {
+      console.error("Error logging out:", error);
+      statusDiv[2].textContent = "An error occurred while logging out.";
+    }
+  });
+
+  //an array to store toggle elements and their checked status for service worker injection
+  const toggles = [
+    { element: socialToggle, checked: false },
```
Comment (udaybansal19) [inline content/content_scripts/popup.js]: we can add an Id here and create switch case in worker using that

Comment (MasterK0927) [reply content/content_scripts/popup.js]: Implementing the same

File: service_worker/background.js [hunk 24]
```diff
@@ -0,0 +1,326 @@
...
+        } else {
+            sendResponse({ success: false, error: '*Invalid Credentials' });
+        }
+    } catch (error) {
+        console.error('Error logging out:', error);
+        sendResponse({ success: false, error: 'An error occurred while logging out' });
+    }
+}
+
+export const injectServiceWorker = async (toggles,checkedToggles) => {
```
Comment (udaybansal19) [inline service_worker/background.js]: toggles is not needed right

**Review by udaybansal19 — APPROVED** [review]
Approving since this has been tested

**Pull request merged**